### PR TITLE
cmake: turn on sysmalloc with ASAN build

### PR DESCRIPTION
**Pull request by mandesero**
This patch enables `LUAJIT_USE_SYSMALLOC` in ASAN build when also GC64 mode is enabled (to avoid allocator and pointer bitness incompatibility).

Closes https://github.com/tarantool/tarantool/issues/10550
Follows up https://github.com/tarantool/tarantool/issues/3071

NO_TEST=internal
NO_DOC=internal
NO_CHANGELOG=internal

### CODE REVIEW

File: cmake/luajit.cmake [hunk 1]
```diff
@@ -69,14 +69,10 @@ endif()
...
-# allocator (i.e. enable LUAJIT_USE_SYSMALLOC option). However,
-# Tarantool doesn't finalize Lua universe the right way (see the
-# comments near <tarantool_lua_free>), so running Tarantool
-# testing routine with LUAJIT_USE_SYSMALLOC enabled generates
-# false-positive LSan leaks. Return back here to enable
-# LUAJIT_USE_SYSMALLOC, when the issue below is resolved.
-# https://github.com/tarantool/tarantool/issues/3071
+# allocator (i.e. enable LUAJIT_USE_SYSMALLOC option).
 if(ENABLE_ASAN)
+    set(LUAJIT_USE_SYSMALLOC ON CACHE BOOL
```
Comment (Buristan) [inline cmake/luajit.cmake:74]: For non-GC64 builds we can't use sysalloc since the bitness of the allocator and pointers should be the same. See [the following build failure](https://github.com/tarantool/tarantool/actions/runs/10793652440/job/29936087657?pr=10554#step:8:3760).

I suppose we should enable this flag only when the `LUAJIT_ENABLE_GC64` is given as well.

**Review by Buristan — CHANGES_REQUESTED** [review]
Hi, Maxim!
Thanks for the patch!

Please reference #10550 to be closed. 
Also, please consider my comment below.

File: cmake/luajit.cmake [hunk 2]
```diff
@@ -69,14 +69,15 @@ endif()
...
-# testing routine with LUAJIT_USE_SYSMALLOC enabled generates
-# false-positive LSan leaks. Return back here to enable
-# LUAJIT_USE_SYSMALLOC, when the issue below is resolved.
-# https://github.com/tarantool/tarantool/issues/3071
+# allocator (i.e. enable LUAJIT_USE_SYSMALLOC option). Also we need
+# to set LUAJIT_ENABLE_GC64 because for non-GC64 builds we can't
+# use sysmalloc since the bitness of the allocator and pointers should
+# be the same.
 if(ENABLE_ASAN)
+    set(LUAJIT_ENABLE_GC64 ON CACHE BOOL
```
Comment (Buristan) [inline cmake/luajit.cmake:77]: I suppose we shouldn't silently enable GC64 when the corresponding flag isn't set.
I think we should enable `LUAJIT_USE_SYSMALLOC` only for GC64 build and raise a _warning_ otherwise.

Comment (mandesero) [reply cmake/luajit.cmake]: Okay, `LUAJIT_USE_SYSMALLOC` is set only when `GC64=ON`.

File: cmake/luajit.cmake [hunk 3]
```diff
@@ -66,17 +66,21 @@ if(ENABLE_VALGRIND)
...
-# comments near <tarantool_lua_free>), so running Tarantool
-# testing routine with LUAJIT_USE_SYSMALLOC enabled generates
-# false-positive LSan leaks. Return back here to enable
-# LUAJIT_USE_SYSMALLOC, when the issue below is resolved.
-# https://github.com/tarantool/tarantool/issues/3071
+# Enable LuaJIT ASan support.
 if(ENABLE_ASAN)
+    if (LUAJIT_ENABLE_GC64)
+        set(LUAJIT_USE_SYSMALLOC ON CACHE BOOL
+            "System provided memory allocator (realloc)" FORCE)
```
Comment (ligurio) [inline cmake/luajit.cmake:73]: I would remove `(realloc)` or replace with `(realloc/malloc)`.

**Review by ligurio — APPROVED** [review]
LGTM after fixing comments from Sergey K.

File: cmake/luajit.cmake (hunk 3 above)
Comment (mandesero) [reply cmake/luajit.cmake]: Replaced with `(realloc/malloc)`.

File: cmake/luajit.cmake [hunk 4]
```diff
@@ -66,17 +66,21 @@ if(ENABLE_VALGRIND)
...
+        set(LUAJIT_USE_SYSMALLOC ON CACHE BOOL
+            "System provided memory allocator (realloc/malloc)" FORCE)
+    else()
+        message(WARNING
+            "The internal LuaJIT memory allocator is not instrumented yet,"
+            " so to find any memory faults, it's worth building LuaJIT with"
+            " the system provided memory allocator. This requires enabling"
+            " the LUAJIT_ENABLE_GC64 option, as sysmalloc can only be used"
+            " in GC64 mode. Without it, sysmalloc cannot be used due to"
+            " allocator and pointer bitness incompatibility"
```
Comment (Buristan) [inline cmake/luajit.cmake:81]: Minor: Missed dot at the end of the sentence.

File: cmake/luajit.cmake [hunk 5]
```diff
@@ -66,17 +66,21 @@ if(ENABLE_VALGRIND)
...
-# allocator (i.e. enable LUAJIT_USE_SYSMALLOC option). However,
-# Tarantool doesn't finalize Lua universe the right way (see the
-# comments near <tarantool_lua_free>), so running Tarantool
-# testing routine with LUAJIT_USE_SYSMALLOC enabled generates
-# false-positive LSan leaks. Return back here to enable
-# LUAJIT_USE_SYSMALLOC, when the issue below is resolved.
-# https://github.com/tarantool/tarantool/issues/3071
+# Enable LuaJIT ASan support.
 if(ENABLE_ASAN)
+    if (LUAJIT_ENABLE_GC64)
```
Comment (Buristan) [inline cmake/luajit.cmake:71]: Typo: s/if (/if(/

**Review by Buristan — CHANGES_REQUESTED** [review]
Hi, Maxim!
Thanks for the fixes!
LGTM, after fixing a couple of nits below.

> This patch enables `LUAJIT_USE_SYSMALLOC` in ASAN build after
resolution of #3071.
> 
> NO_TEST=internal
NO_DOC=internal
NO_CHANGELOG=internal

I would rephrase it like the following (to accent that this is done only for GC64 mode):

> This patch enables `LUAJIT_USE_SYSMALLOC` in ASAN build when also GC64 mode is enabled (to avoid allocator and pointer bitness incompatibility). 
> 
> Closes #10550
> Follows up #3071

File: cmake/luajit.cmake (hunk 4 above)
Comment (mandesero) [reply cmake/luajit.cmake]: Fixed.

File: cmake/luajit.cmake (hunk 5 above)
Comment (mandesero) [reply cmake/luajit.cmake]: Fixed.

**Review by Buristan — APPROVED** [review]

File: cmake/luajit.cmake (hunk 3 above)
Comment (ligurio) [reply cmake/luajit.cmake]: Thanks!
